indexer.ts

```diff
@@ -0,0 +1,743 @@
+  }
+
+  // helper function to format
+  function formatToNear(yoctoAmount) {
+    const nearAmount = yoctoAmount / 10 ** 24;
+    return nearAmount;
+  }
+
+  async function handleNewDonations(
+    args,
+    receiverId,
+    signerId,
+    actionName,
+    receiptId
+  ) {
+    let matching_pool = JSON.parse(args).matching_pool
+    if ((actionName == "donate" && receiverId != "donate.potlock.near") || matching_pool) {
+      console.log("calling donate on projects...");
+      return;
+    }
+    console.log("action and recv", actionName, receiverId);
+    let receipt = block
+      .receipts()
+      .filter((receipt) => receipt.receiptId == receiptId)[0];
+    let donation_data = JSON.parse(atob(receipt.status["SuccessValue"]));
+    console.log("result arg...:", donation_data);
+    let donated_at = new Date(donation_data.donated_at || donation_data.donated_at_ms);
+    await context.db.Account.upsert({ id: donation_data.donor_id }, ["id"], []);
+    let endpoint = `${GECKO_URL}/coins/${
+      donation_data.ft_id || "near"
+    }/history?date=${formatDate(donated_at)}&localization=false`;
+    let response = await fetch(endpoint);
+    let data = await response.json();
+    let unit_price = data.market_data?.current_price.usd;
+    if (!unit_price) {
+      console.log("api rate limi?::", data)
+      unit_price = parseFloat(localStorage.getItem("last_price"))
+    }
+    localStorage.setItem("last_price", unit_price.toString()) // store last price incase subsequent price fetch fails,  does it support local storage?
+
+    let total_amount = donation_data.total_amount;
+    let net_amount = (
+      BigInt(donation_data.total_amount) - BigInt(donation_data.protocol_fee)
+    ).toString();
+
+    let totalnearAMount = formatToNear(total_amount);
+    let netnearAMount = formatToNear(net_amount);
+    let total_amount_usd = unit_price * totalnearAMount
+    let net_amount_usd = unit_price * netnearAMount
+    let donation = {
+      donor_id: donation_data.donor_id,
+      total_amount,
+      total_amount_usd,
+      net_amount_usd,
+      net_amount,
+      ft_id: donation_data.ft_id || null,
+      message: donation_data.message,
+      donated_at,
+      matching_pool: donation_data.matching_pool || false,
+      recipient_id: donation_data.project_id || donation_data.recipient_id,
+      protocol_fee: donation_data.protocol_fee,
+      referrer_id: donation_data.referrer_id,
+      referrer_fee: donation_data.referrer_fee,
+      tx_hash: receiptId,
+    };
+    await context.db.Donation.insert(donation);
+
+    if (actionName != "donate") { // update pot data such as public donations and matchingpool
+      let pot = (await context.db.Pot.select({ id: receiverId}))[0]
+      donation["pot_id"] = pot.id
+      let potUpdate = {
+        total_public_donations_usd: pot.total_public_donations_usd || 0 + total_amount_usd,
+        total_public_donations: pot.total_public_donations || 0 + total_amount,
```

Comment:
why do you need the `or` statement? aren't these `NOT NULL` fields?

Reply:
for first time entries, since it's null at the start.